Add docs for onCaughtError and onUncaughtError (#6742)
reactjs/react.dev: src/content/reference/react-dom/client/createRoot.md

* **optional** `options`: An object with options for this React root.

  * <CanaryBadge title="This feature is only available in the Canary channel" /> **optional** `onCaughtError`: Callback called when React catches an error in an error boundary. Called with the `error` and an `errorInfo` object containing the `componentStack`.

>   * <CanaryBadge title="This feature is only available in the Canary channel" /> **optional** `onUncaughtError`: Callback called when an uncaught error is thrown. Called with the `error` and an `errorInfo` object containing the `componentStack`.

Reviewer: Incoming "well ackshually": "Throwing an uncaught error" is technically not really a thing. You don't know yet if an error is caught or not when you throw it.

```suggestion
  * <CanaryBadge title="This feature is only available in the Canary channel" /> **optional** `onUncaughtError`: Callback called when a thrown error is not caught. Called with the `error` and an `errorInfo` object containing the `componentStack`.
```

Author: well ackshually, react catches the error in the component, and rethrows the error if it's uncaught by an error bounday, so _technically_ react re-throws the uncaught error. But from the user's perspective you're right, it should be "when a thrown error is not caught" 😜

Author: Fixed. I also update the descriptions to say:
- uncaught: Called with the `error` that was thrown
- caught: Called with the `error` caught by the Error Boundary
- recoverable: Called with the `error` React throws

>   * **optional** `onRecoverableError`: Callback called when React automatically recovers from errors. Called with the `error` and an `errorInfo` object containing the `componentStack`. Some recoverable errors may include the original error cause as `error.cause`.

Reviewer: Can we link to a section explaining what recovering from an error means in React?

It is uncommon to call `render` multiple times. Usually, your components will [update state](/reference/react/useState) instead.

### Show a dialog for uncaught errors {/*show-a-dialog-for-uncaught-errors*/}

<Canary>

`onUncaughtError` is only available in the latest React Canary release.

</Canary>

> By default, React will log all uncaught errors caught to the console. To implement your own error reporting, you can provide the optional `onUncaughtError` root option:

Reviewer: ```suggestion
By default, React will log all uncaught errors to the console. To implement your own error reporting, you can provide the optional `onUncaughtError` root option:
```

```js [[1, 6, "onUncaughtError"], [2, 6, "error", 1], [3, 6, "errorInfo"], [4, 10, "componentStack"]]
import { createRoot } from 'react-dom/client';

const root = createRoot(
  document.getElementById('root'),
  {
    onUncaughtError: (error, errorInfo) => {
      console.error(
        'Uncaught error',
        error,
        errorInfo.componentStack
      );
    }
  }
);
root.render(<App />);
```

The <CodeStep step={1}>onUncaughtError</CodeStep> option is a function called with two arguments:

> 1. The <CodeStep step={2}>error</CodeStep> that was caught.

Reviewer: May be confusing that for both "on uncaught" and "on caught" error we describe the error as the "caught" error? 
```suggestion
1. The <CodeStep step={2}>error</CodeStep> that was not caught.
```

Author: Good point, I updated it to say:
- uncaught: the error that was thrown
- caught: the error that was caught by the boundary
- recoverable: the error React throws

  reportError({ title: "Caught Error", error, componentStack,  dismissable: true});
}

export function reportUncaughtError({error, cause, componentStack}) {
  reportError({ title: "Uncaught Error", error, componentStack, dismissable: false });
}

export function reportRecoverableError({error, cause, componentStack}) {
  reportError({ title: "Recoverable Error", error, componentStack,  dismissable: true });
}
```

```js src/index.js active
import { createRoot } from "react-dom/client";
import App from "./App.js";
import {reportUncaughtError} from "./reportError";
import "./styles.css";

const container = document.getElementById("root");
const root = createRoot(container, {
  onUncaughtError: (error, errorInfo) => {
    if (error.message !== 'Known error') {
      reportUncaughtError({
        error,
        componentStack: errorInfo.componentStack
      });
    }
  }
});
root.render(<App />);
```

```js src/App.js
import { useState } from 'react';
import { ErrorBoundary } from "react-error-boundary";

export default function App() {
  const [throwError, setThrowError] = useState(false);
  
  if (throwError) {
    foo.bar = 'baz';
  }
  
  return (
    <>

>       <ErrorBoundary

Reviewer 2: What's the benefit of having an ErrorBoundary here? Would be a simpler example without one and less confusing when comparing to `onCaughtError`. I at least wondered why this wasn't being caught before reading the code to see the state change triggers an error above the boundary. Unless the point is to have people think through that?

Author: Great point, updated.

        fallbackRender={fallbackRender}
        onReset={(details) => {
          setThrowError(false);
        }}
      >

>         <span>This error show the error dialog:</span>

Reviewer 2: ```suggestion
        <span>This error shows the error dialog:</span>
```

Author: Fixed

        <button onClick={() => setThrowError(true)}>
          Throw error
        </button>
      </ErrorBoundary>
      
    </>
  );
}

function fallbackRender({ resetErrorBoundary }) {
  return (
    <div role="alert">
      <h3>Error Boundary</h3>
      <p>Something went wrong.</p>
      <button onClick={resetErrorBoundary}>Reset</button>
    </div>
  );
}
```

```json package.json hidden
{
  "dependencies": {
    "react": "canary",
    "react-dom": "canary",
    "react-scripts": "^5.0.0",
    "react-error-boundary": "4.0.3"
  },
  "main": "/index.js"
}
```

</Sandpack>

### Displaying error boundary errors {/*displaying-error-boundary-errors*/}

<Canary>

`onCaughtError` is only available in the latest React Canary release.

</Canary>

> By default, React will log all errors caught by an error boundary to `console.error`. To override this behavior, you can provide the optional `onCaughtError` root option to handle errors caught by an [Error Boundary](/reference/react/Component#catching-rendering-errors-with-an-error-boundary):

Reviewer: Is Error Boundary a React concept? I feel like it is even though we don't have a good page explaining it. There are more instances in this PR where we use the lowercase "error boundary.

Author: Yes, we have docs [here](https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary) (and I link to it elsewhere on the page)

I'll fix the capitalization in the files edited here, but we have https://github.com/reactjs/react.dev/issues/6713 open to address the capitalizion in other docs
